orangecontrib/geo/widgets/owmap.py

```diff
@@ -1,906 +1,849 @@
-
-    class Warning(widget.OWWidget.Warning):
-        all_nan_slice = widget.Msg('Latitude and/or longitude has no defined values (is all-NaN)')
-
-    UserAdviceMessages = [
-        widget.Message(
-            'Select markers by holding <b><kbd>Shift</kbd></b> key and dragging '
-            'a rectangle around them. Clear the selection by clicking anywhere.',
-            'shift-selection')
-    ]
+    settings_version = 3
+    tile_provider = settings.Setting(next(iter(TILE_PROVIDERS)))
+    attr_lat = settings.ContextSetting(None)
+    attr_lon = settings.ContextSetting(None)
+    tooltip_shows_all = settings.Setting(True)
+
+    GRAPH_CLASS = OWScatterPlotMapGraph
+    graph = settings.SettingProvider(OWScatterPlotMapGraph)
+    embedding_variables_names = None
+
+    class Warning(OWDataProjectionWidget.Warning):
+        missing_coords = Msg(
+            "Plot cannot be displayed because '{}' or '{}' "
+            "is missing for all data points")
+        no_lat_lon_vars = Msg("Data has no latitude and longitude variables.")
+        out_of_range = Msg("Latitude or longitude is out of range.")
+        no_internet = Msg("Cannot fetch map from the internet. "
+                          "Displaying only cached parts.")
+
+    class Information(OWDataProjectionWidget.Information):
+        missing_coords = Msg(
+            "Points with missing '{}' or '{}' are not displayed")
+
+    def _add_controls(self):
+        self.lat_lon_model = DomainModel(DomainModel.MIXED,
+                                         valid_types=ContinuousVariable)
+
+        lat_lon_box = gui.vBox(self.controlArea, True)
+        options = dict(
+            labelWidth=50, orientation=Qt.Horizontal, sendSelectedValue=True,
+            valueType=str, contentsLength=14
+        )
+
+        gui.comboBox(lat_lon_box, self, 'tile_provider', label='Map:',
+                     items=list(TILE_PROVIDERS.keys()),
+                     callback=self._set_tile_provider, **options)
+
+        gui.comboBox(lat_lon_box, self, 'attr_lat', label='Lat:',
+                     callback=self.set_attr_from_combo,
+                     model=self.lat_lon_model, **options)
+
+        gui.comboBox(lat_lon_box, self, 'attr_lon', label='Lon:',
+                     callback=self.set_attr_from_combo,
+                     model=self.lat_lon_model, **options)
+
+        super()._add_controls()
+
+        gui.checkBox(
+            self._plot_box, self,
+            value="graph.freeze",
+            label="Freeze map",
+            tooltip="If checked, the map won't change position to fit new data.")
+
+    def _set_tile_provider(self):
+        self.graph.clear_map()
+        self.graph.set_tile_provider(TILE_PROVIDERS[self.tile_provider])
+        self.graph.update_map()
+
+    def check_data(self):
+        super().check_data()
+        if self.data is not None:
+            if not self.data.domain.has_continuous_attributes(True, True):
```

Comment:
It would be useful to include a warning here. Something like "`Data has no continuous variables."`.